pkg/record/record.go

```diff
@@ -0,0 +1,71 @@
+package record
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"io"
+)
+
+// ErrIllegalTopicName is returned if a topic's character sequence is invalid.
+var ErrIllegalTopicName = errors.New("illegal topic name")
+
+// Reader emits records.
+// It returns io.EOF if the underlying record source has no more data.
+type Reader func() (record []byte, err error)
+
+// RecordReaderFactory returns a new RecordReader backed by the given reader.
+type ReaderFactory func(io.Reader) Reader
+
+// NewDynamicReader returns a record reader that expects each input record from r
+// to start with a space-delimited topic identifier.
+func NewDynamicReader(r io.Reader) Reader {
+	br := bufio.NewReader(r)
+
+	return func() ([]byte, error) {
+		l, err := br.ReadBytes('\n')
+		if err != nil {
+			return nil, err
+		}
+		// Validate that a correct topic prefix exists.
+		if i := bytes.IndexByte(l, ' '); i < 0 || !IsValidTopic(l[:i]) {
+			return nil, ErrIllegalTopicName
+		}
+		return l, nil
+	}
+}
+
+// StaticReaderFactory returns a RecordReaderFactory that prefixes all records
+// with the given bytes as topic.
+// Callers must ensure the topic identifier is valid.
+func StaticReaderFactory(topic []byte) ReaderFactory {
+	topic = append(topic, ' ')
```

Comment:
Couldn't we use `IsValidTopic` for validation in this constructor and return an error accordingly?

Reply:
I think we had that initially but I removed it per the signature suggestion from Peter above. Can add it back in case it wasn't dropped there intentionally.

Or we rather just name the argument `prefix` in this case and adjust docs so it's clear validation is the callers responsibility.